src/Storages/MergeTree/MergeTreeIndexGin.cpp

```diff
@@ -0,0 +1,786 @@
+
+    const auto & size_type = DataTypePtr(std::make_shared<DataTypeUInt32>());
+    auto size_serialization = size_type->getDefaultSerialization();
+
+    for (const auto & gin_filter : gin_filters)
+    {
+        size_t filter_size = gin_filter.getFilter().size();
+        size_serialization->serializeBinary(filter_size, ostr);
+        ostr.write(reinterpret_cast<const char*>(gin_filter.getFilter().data()), filter_size * sizeof(RowIDRange));
+    }
+}
+
+void MergeTreeIndexGranuleGinFilter::deserializeBinary(ReadBuffer & istr, MergeTreeIndexVersion version)
+{
+    if (version != 1)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Unknown index version {}.", version);
+
+    Field field_rows;
+    const auto & size_type = DataTypePtr(std::make_shared<DataTypeUInt32>());
+
+    for (auto & gin_filter : gin_filters)
+    {
+        size_type->getDefaultSerialization()->deserializeBinary(field_rows, istr);
+        size_t filter_size = field_rows.get<size_t>();
+
+        if (filter_size == 0)
+            continue;
+
+        gin_filter.getFilter().assign(filter_size, {});
+        istr.read(reinterpret_cast<char*>(gin_filter.getFilter().data()), filter_size * sizeof(RowIDRange));
+    }
+    has_elems = true;
+}
+
+
+MergeTreeIndexAggregatorGinFilter::MergeTreeIndexAggregatorGinFilter(
+    GinIndexStorePtr store_,
+    const Names & index_columns_,
+    const String & index_name_,
+    const GinFilterParameters & params_,
+    TokenExtractorPtr token_extractor_)
+    : store(store_)
+    , index_columns(index_columns_)
+    , index_name (index_name_)
+    , params(params_)
+    , token_extractor(token_extractor_)
+    , granule(
+        std::make_shared<MergeTreeIndexGranuleGinFilter>(
+            index_name, index_columns.size(), params))
+{
+}
+
+MergeTreeIndexGranulePtr MergeTreeIndexAggregatorGinFilter::getGranuleAndReset()
+{
+    auto new_granule = std::make_shared<MergeTreeIndexGranuleGinFilter>(
+        index_name, index_columns.size(), params);
+    new_granule.swap(granule);
+    return new_granule;
+}
+
+void MergeTreeIndexAggregatorGinFilter::addToGinFilter(UInt32 rowID, const char* data, size_t length, GinFilter& gin_filter)
+{
+    size_t cur = 0;
+    size_t token_start = 0;
+    size_t token_len = 0;
+
+    while (cur < length && token_extractor->nextInStringPadded(data, length, &cur, &token_start, &token_len))
+    {
+        gin_filter.add(data + token_start, token_len, rowID, store);
+    }
+}
+
+void MergeTreeIndexAggregatorGinFilter::update(const Block & block, size_t * pos, size_t limit)
+{
+    if (*pos >= block.rows())
+        throw Exception(
+                "The provided position is not less than the number of block rows. Position: "
+                + toString(*pos) + ", Block rows: " + toString(block.rows()) + ".", ErrorCodes::LOGICAL_ERROR);
+
+    size_t rows_read = std::min(limit, block.rows() - *pos);
+    auto row_id = store->getNextRowIDRange(rows_read);
```

Comment:
This returns UInt32, how sure are you of such limit? I wouldn't be so. 4 billion rows and I believe you have an overlap

Reply:
RowID range is defined within a segment(default is 256M), so one segment usually doesn't have more than 4 billion rows. Even for one part, the size is usually about hundreds of GB (suppose each row has 100 bytes, for example) for 4 billion of rows which would be split into more parts in ClickHouse.
So using 64-bit row ID is overkill.